### ISSUE TITLE
Show number of times `coder_app` is opened in Template Insights

### ISSUE THREAD

**Issue by stirby** [coder/coder#13099]
We expose the number of hours each coder_app and IDE is used for a given interval, but not the number of distinct times it was used.

We should add a count for the number of times it was opened. 

This will help identify the frequency of app usage via click ops. 

Added some design mockups for inspiration:
![Screenshot 2024-04-29 at 11 54 57 AM](https://github.com/coder/coder/assets/58410745/320a3d7f-6546-46a0-8352-93a59538432a)

**Comment by stirby** [comment]
Also could use "visits" or "sessions".

**Comment by bpmct** [comment]
huge fan of this!

**Comment by mafredri** [comment]
I can see that this would be interesting data to show. However, it raises the question for how do we define number of times opened?

If we take code-server, for instance. A reload of the app will produce a bunch of requests in short succession, and leave some long-running ones as well (WebSocket). This property is alright when we’re only looking at the boolean, did/did not use. But if we want an accurate count, we need to know what/how to count. Behavior will also differ between different coder apps, so perhaps a template author will need a way to define this.

**Comment by stirby** [comment]
Good point @mafredri. 

What would be most general is to denote "sessions" just to be on the same page for terminology.  

If there's a way to distinguish the click-op of the opening an app via the UI (when there's no existing usage) that would be easiest. If the time series data is already accessible we could use some sensible default for congruent sessions and maybe allow admins to configure it in the `coder_app` resource? 

Do all `coder_app`s have these long-running websocket connections?

I think it's most dependent on the accessibility of time-series usage per workspace, per app.

**Comment by stirby** [comment]
Oh great, thank you for all the detail here @mafredri. 


> We list "clicks" only via the channels we're able to: dashboard (and potentially CLI in future)
This is definitely the simplest way to satisfy the current customer requests. Ben has recently remarked that he'd like to expand the capibilites of `coder open ...`; I am aligned with his interest and appreciate you keeping that in mind in designing this (although changing that is out of scope for this). 


> 2. We count "opened N times" via session length and a timeout. Let's say we define this with a 30 minutes timeout, if the user made a request every minute for 18 minutes, then no requests for the following 30 minutes, we would count this as "opened 1 time, duration 18 minutes".

I would prefer this because our customers are wanting to learn about the frequency of usage for IDEs/Tools which will always need bi-directional connections, and it captures finer grained detail. I think customers will be okay with an approximation since they'll be looking at this more per-group or per-template and want broad-stroke aggregations.

> We could analyze existing app usage on our dogfood instance to determine a good timeout, but the usefulness is limited due to there not being much usage outside of code-server (unverified claim).

If it's low-lift, I think we should test both 1 and 2 out in dogfood, there's enough app usage for some rough tuning:
![Screenshot 2024-05-14 at 2 35 21 PM](https://github.com/coder/coder/assets/58410745/88d8ba26-0d79-4fa0-88ab-f2b13e283fe1)

Otherwise we should aim for #2.

**Comment by mafredri** [comment]
@stirby thanks, that helps clarify this task. I think we should table 1. if we feel 2. is sufficient because I can think of two options for tracking this, where one approach could be considered in a greater whole wrt analytics of dashboard/CLI usage. Since 1. is new data, we also need to a place to store it in the database.

Sans DB modifications, I'd say 1. is low-lift if we consider it in isolation, though.

**Comment by mafredri** [comment]
@stirby Once #13292 is merged, the API will include `times_used` for `coder_app`s only (e.g. code-server, **not** VS Code, SSH, etc.). I'll leave it up to someone else how we would present this information in the dashboard, and if we should put it behind an experiment.

This is based on option 2, i.e. an approximate count (30m-1h of idle time results in a new usage count the next time it's used).

**Comment by mtojek** [comment]
> I'll leave it up to someone else how we would present this information in the dashboard, and if we should put it behind an experiment.

@BrunoQuaresma Is is something you can help with?

**Comment by BrunoQuaresma** [comment]
@mtojek sure thing.

**Comment by BrunoQuaresma** [comment]
@stirby wdyt about the following design?

<img width="890" alt="Screenshot 2024-05-21 at 14 14 37" src="https://github.com/coder/coder/assets/3165839/803142da-362d-420a-ae0f-aa6de623b253">

**Comment by stirby** [comment]
Phenomenal.

**Comment by michaelbrewer** [comment]
Awesome. Will this be in the next mainline update ?

**Comment by matifali** [comment]
@michaelbrewer yes.
